builtin/providers/aws/auth_helpers.go

```diff
@@ -0,0 +1,170 @@
+		}
+
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return "", fmt.Errorf("Failed reading EC2 metadata API response from %q: %s", url, err)
+		}
+		info := IamInfoFromMetadata{}
+		err = json.Unmarshal(body, &info)
+		if err != nil {
+			return "", fmt.Errorf("Failed parsing EC2 metadata API response from %q: %s", url, err)
+		}
+
+		return parseAccountIdFromArn(info.InstanceProfileArn)
+	}
+
+	// Then try IAM GetUser
+	log.Println("[DEBUG] Trying to get account ID via iam:GetUser")
+	outUser, err := iamconn.GetUser(nil)
+	if err == nil {
+		return parseAccountIdFromArn(*outUser.User.Arn)
+	}
+
+	// Then try IAM ListRoles
+	awsErr, ok := err.(awserr.Error)
+	// AccessDenied and ValidationError can be raised
+	// if credentials belong to federated profile, so we ignore these
+	if !ok || (awsErr.Code() != "AccessDenied" && awsErr.Code() != "ValidationError") {
+		return "", fmt.Errorf("Failed getting account ID via 'iam:GetUser': %s", err)
+	}
+
+	log.Printf("[DEBUG] Getting account ID via iam:GetUser failed: %s", err)
+	log.Println("[DEBUG] Trying to get account ID via iam:ListRoles instead")
+	outRoles, err := iamconn.ListRoles(&iam.ListRolesInput{
+		MaxItems: aws.Int64(int64(1)),
+	})
+	if err != nil {
+		return "", fmt.Errorf("Failed getting account ID via 'iam:ListRoles': %s", err)
+	}
+
+	if len(outRoles.Roles) < 1 {
+		return "", fmt.Errorf("Failed getting account ID via 'iam:ListRoles': No roles available")
+	}
+
+	return parseAccountIdFromArn(*outRoles.Roles[0].Arn)
+}
+
+func parseAccountIdFromArn(arn string) (string, error) {
+	parts := strings.Split(arn, ":")
+	if len(parts) < 5 {
+		return "", fmt.Errorf("Unable to parse ID from invalid ARN: %q", arn)
+	}
+	return parts[4], nil
+}
+
+// AWS will add a "Server: EC2" header value to the metadata response.
+// This allows checking the headers for this value to ensure
+// something else didn't just happen to be listening on that IP:Port
+func responseContainsEC2Header(resp *http.Response) bool {
+	return resp.Header["Server"] != nil && strings.Contains(resp.Header["Server"][0], "EC2")
+}
+
+// This function is responsible for reading credentials from the
+// environment in the case that they're not explicitly specified
+// in the Terraform configuration.
+func getCreds(key, secret, token, profile, credsfile string) *awsCredentials.Credentials {
+	// build a chain provider, lazy-evaulated by aws-sdk
+	providers := []awsCredentials.Provider{
+		&awsCredentials.StaticProvider{Value: awsCredentials.Value{
+			AccessKeyID:     key,
+			SecretAccessKey: secret,
+			SessionToken:    token,
+		}},
+		&awsCredentials.EnvProvider{},
+		&awsCredentials.SharedCredentialsProvider{
+			Filename: credsfile,
+			Profile:  profile,
+		},
+	}
+
+	// We only look in the EC2 metadata API if we can connect
+	// to the metadata service within a reasonable amount of time
+	metadataURL := buildMetadataUrl("/latest")
```

Comment:
That's not a bad idea, but we'll have to check what's the default timeout. I assume there was a reason to add this "workaround" instead of leaving the EC2Role provider in the chain for auto-detection.